src/placeholder.ts

```diff
@@ -0,0 +1,422 @@
+import {Parser, tokTypes} from "acorn";
+import {acornOptions} from "./javascript/parse.js";
+
+const CODE_TAB = 9,
+  CODE_LF = 10,
+  CODE_FF = 12,
+  CODE_CR = 13,
+  CODE_SPACE = 32,
+  CODE_UPPER_A = 65,
+  CODE_UPPER_Z = 90,
+  CODE_LOWER_A = 97,
+  CODE_LOWER_Z = 122,
+  CODE_LT = 60,
+  CODE_GT = 62,
+  CODE_SLASH = 47,
+  CODE_DASH = 45,
+  CODE_BANG = 33,
+  CODE_EQ = 61,
+  CODE_DQUOTE = 34,
+  CODE_SQUOTE = 39,
+  CODE_QUESTION = 63,
+  CODE_DOLLAR = 36,
+  CODE_LBRACE = 123,
+  CODE_RBRACE = 125,
+  CODE_BACKSLASH = 92,
+  STATE_DATA = 1,
+  STATE_TAG_OPEN = 2,
+  STATE_END_TAG_OPEN = 3,
+  STATE_TAG_NAME = 4,
+  STATE_BOGUS_COMMENT = 5,
+  STATE_BEFORE_ATTRIBUTE_NAME = 6,
+  STATE_AFTER_ATTRIBUTE_NAME = 7,
+  STATE_ATTRIBUTE_NAME = 8,
+  STATE_BEFORE_ATTRIBUTE_VALUE = 9,
+  STATE_ATTRIBUTE_VALUE_DOUBLE_QUOTED = 10,
+  STATE_ATTRIBUTE_VALUE_SINGLE_QUOTED = 11,
+  STATE_ATTRIBUTE_VALUE_UNQUOTED = 12,
+  STATE_AFTER_ATTRIBUTE_VALUE_QUOTED = 13,
+  STATE_SELF_CLOSING_START_TAG = 14,
+  STATE_COMMENT_START = 15,
+  STATE_COMMENT_START_DASH = 16,
+  STATE_COMMENT = 17,
+  STATE_COMMENT_LESS_THAN_SIGN = 18,
+  STATE_COMMENT_LESS_THAN_SIGN_BANG = 19,
+  STATE_COMMENT_LESS_THAN_SIGN_BANG_DASH = 20,
+  STATE_COMMENT_LESS_THAN_SIGN_BANG_DASH_DASH = 21,
+  STATE_COMMENT_END_DASH = 22,
+  STATE_COMMENT_END = 23,
+  STATE_COMMENT_END_BANG = 24,
+  STATE_MARKUP_DECLARATION_OPEN = 25,
+  STATE_RAWTEXT = 26,
+  STATE_RAWTEXT_LESS_THAN_SIGN = 27,
+  STATE_RAWTEXT_END_TAG_OPEN = 28,
+  STATE_RAWTEXT_END_TAG_NAME = 29;
+
+export interface PlaceholderToken {
+  type: "content" | "code";
```

Comment:
I expect it would now be possible to interpolate into attributes (like Hypertext Literal does). However, this will require a specialized version of the display function (that coerces to a string). I’m leaving this as future work but we are now much closer to fixing #32.